plugin.program.remote.control.browser/plugin.py

```diff
@@ -0,0 +1,990 @@
+        xbmc.log('Not searching for process descendents', xbmc.LOGDEBUG)
+        return [parent]
+    try:
+        process = psutil.Process(parent)
+        try:
+            children = process.children
+        except AttributeError:
+            children = process.get_children
+        processes = [process] + children(recursive=True)
+        return [node.pid for node in processes]
+    except psutil.NoSuchProcess:
+        xbmc.log('Failed to find process tree', xbmc.LOGDEBUG)
+        return []
+
+
+def killBrowser(proc, sig):
+    for pid in getProcessTree(proc.pid):
+        try:
+            os.kill(pid, sig)
+        except OSError:
+            pass
+
+
+@contextlib.contextmanager
+def runBrowser(browserCmd):
+    (sink, source) = socket.socketpair()
+    with contextlib.closing(sink), contextlib.closing(source):
+        waiter = None
+        try:
+            if xbmc.getCondVisibility('System.Platform.Windows'):
+                # On Windows, the Popen will block unless close_fds is True and
+                # creationflags is DETACHED_PROCESS.
+                xbmc.log('Using Windows creation flags', xbmc.LOGDEBUG)
+                creationflags = 0x00000008
+            else:
+                creationflags = 0
+            xbmc.log(
+                'Launching browser: ' +
+                ' '.join(pipes.quote(arg) for arg in browserCmd),
+                xbmc.LOGINFO)
+            proc = subprocess.Popen(
+                browserCmd, creationflags=creationflags, close_fds=True)
+            try:
+                # Monitor the browser and kick the socket when it exits.
+                waiterStarting = threading.Thread(
+                    target=monitorProcess, args=(proc, source))
+                waiterStarting.start()
+                waiter = waiterStarting
+
+                yield (proc, sink.fileno())
+
+                # Ask each child process to exit.
+                xbmc.log('Terminating the browser', xbmc.LOGDEBUG)
+                killBrowser(proc, signal.SIGTERM)
+
+                # Give the browser a few seconds to shut down gracefully.
+                def terminateBrowser():
+                    xbmc.log(
+                        'Forcefully killing the browser at the deadline',
+                        xbmc.LOGINFO)
+                    killBrowser(proc, signal.SIGKILL)
+                terminator = threading.Timer(
+                    BROWSER_EXIT_DELAY.total_seconds(), terminateBrowser)
+                terminator.start()
+                try:
+                    proc.wait()
+                    proc = None
+                    xbmc.log('Waited for the browser to quit', xbmc.LOGDEBUG)
+                finally:
+                    terminator.cancel()
+                    terminator.join()
+            finally:
+                if proc is not None:
+                    # As a last resort, forcibly kill the browser.
+                    xbmc.log('Forcefully killing the browser', xbmc.LOGINFO)
```

Comment:
Isn't LOGINFO used for extreme conditions that the user would want to be aware of?